File: internal/protocol/protocol_test.go
package protocol

import (
	"bytes"
	"crypto/rand"
	"encoding/binary"
	"encoding/json"
	"testing"
)

func TestEncodeFrame_HttpRequest(t *testing.T) {
	t.Parallel()

	header := &HttpRequestHeader{
		MessageHeader: MessageHeader{Type: "http_request", ID: "req-1"},
		Method:        "POST",
		Host:          "api.my-homelab.kagami.dev",
		Path:          "/users",
		Query:         "page=2&limit=10",
		Headers:       map[string][]string{"Content-Type": {"application/json"}},
	}
	body := []byte(`{"name":"alice"}`)

	frame, err := EncodeFrame(header, body)
	if err != nil {
		t.Fatalf("EncodeFrame: %v", err)
	}

	// Verify frame structure: 4-byte length prefix + JSON header + body.
	if len(frame) < 4 {
		t.Fatal("frame shorter than 4 bytes")
	}

	headerLen := binary.BigEndian.Uint32(frame[:4])
	headerJSON := frame[4 : 4+headerLen]
	bodyBytes := frame[4+headerLen:]

	// Header length matches actual JSON bytes.
	var decoded HttpRequestHeader
	if err := json.Unmarshal(headerJSON, &decoded); err != nil {
		t.Fatalf("unmarshal header: %v", err)
	}
	if decoded.Type != "http_request" {
		t.Errorf("type = %q, want %q", decoded.Type, "http_request")
	}
	if decoded.ID != "req-1" {
		t.Errorf("id = %q, want %q", decoded.ID, "req-1")
	}
	if decoded.Method != "POST" {
		t.Errorf("method = %q, want %q", decoded.Method, "POST")
	}
	if decoded.Host != "api.my-homelab.kagami.dev" {
		t.Errorf("host = %q, want %q", decoded.Host, "api.my-homelab.kagami.dev")
	}
	if decoded.Path != "/users" {
		t.Errorf("path = %q, want %q", decoded.Path, "/users")
	}
	if decoded.Query != "page=2&limit=10" {
		t.Errorf("query = %q, want %q", decoded.Query, "page=2&limit=10")
	}

	// Body preserved.
	if !bytes.Equal(bodyBytes, body) {
		t.Errorf("body = %q, want %q", bodyBytes, body)
	}
}

func TestDecodeFrame_Valid(t *testing.T) {
	t.Parallel()

	header := &HttpRequestHeader{
		MessageHeader: MessageHeader{Type: "http_request", ID: "req-2"},
		Method:        "GET",
		Host:          "test.kagami.dev",
		Path:          "/health",
		Query:         "",
		Headers:       map[string][]string{},
	}
	body := []byte("hello world")

	encoded, err := EncodeFrame(header, body)
	if err != nil {
		t.Fatalf("EncodeFrame: %v", err)
	}

	f, err := DecodeFrame(encoded)
	if err != nil {
		t.Fatalf("DecodeFrame: %v", err)
	}

	var decoded HttpRequestHeader
	if err := json.Unmarshal(f.Header, &decoded); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if decoded.Type != "http_request" {
		t.Errorf("type = %q, want %q", decoded.Type, "http_request")
	}
	if decoded.ID != "req-2" {
		t.Errorf("id = %q, want %q", decoded.ID, "req-2")
	}
	if !bytes.Equal(f.Body, body) {
		t.Errorf("body = %q, want %q", f.Body, body)
	}
}

func TestDecodeFrame_UnknownType(t *testing.T) {
	t.Parallel()

	headerJSON := []byte(`{"type":"unknown_msg","id":"x"}`)
	frame := make([]byte, 4+len(headerJSON))
	binary.BigEndian.PutUint32(frame[:4], uint32(len(headerJSON)))
	copy(frame[4:], headerJSON)

	f, err := DecodeFrame(frame)
	if err != nil {
		t.Fatalf("DecodeFrame should succeed; parsing type is separate: %v", err)
	}

	_, err = ParseHeaderType(f.Header)
	if err == nil {
		t.Fatal("ParseHeaderType should return error for unknown type")
	}
}

func TestRoundTrip_BinaryBody(t *testing.T) {
	t.Parallel()

	// Generate random binary data.
	original := make([]byte, 1024)
	if _, err := rand.Read(original); err != nil {
		t.Fatalf("rand.Read: %v", err)
	}

	header := &HttpResponseHeader{
		MessageHeader: MessageHeader{Type: "http_response", ID: "resp-1"},
		Status:        200,
		Headers:       map[string][]string{"Content-Type": {"application/octet-stream"}},
	}

	encoded, err := EncodeFrame(header, original)
	if err != nil {
		t.Fatalf("EncodeFrame: %v", err)
	}

	f, err := DecodeFrame(encoded)
	if err != nil {
		t.Fatalf("DecodeFrame: %v", err)
	}

	if !bytes.Equal(f.Body, original) {
		t.Error("round-tripped binary body does not match original")
	}
}

func TestControlMessages_NoBody(t *testing.T) {
	t.Parallel()

	tests := []struct {
		name   string
		header any
	}{
		{
			name:   "ping",
			header: &PingHeader{MessageHeader: MessageHeader{Type: "ping", ID: "p-1"}},
		},
		{
			name:   "pong",
			header: &PongHeader{MessageHeader: MessageHeader{Type: "pong", ID: "p-1"}},
		},
		{
			name: "error",
			header: &ErrorHeader{
				MessageHeader: MessageHeader{Type: "error", ID: "e-1"},
				Code:          "timeout",
				Message:       "request timed out",
			},
		},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()

			encoded, err := EncodeFrame(tc.header, nil)
			if err != nil {
				t.Fatalf("EncodeFrame: %v", err)
			}

			f, err := DecodeFrame(encoded)
			if err != nil {
				t.Fatalf("DecodeFrame: %v", err)
			}

			if f.Body != nil {
				t.Errorf("expected nil body for control message, got %d bytes", len(f.Body))
			}

			msgType, err := ParseHeaderType(f.Header)
			if err != nil {
				t.Fatalf("ParseHeaderType: %v", err)
			}
			if msgType != tc.name {
				t.Errorf("type = %q, want %q", msgType, tc.name)
			}
		})
	}
}

func TestDecodeFrame_ZeroLengthBody(t *testing.T) {
	t.Parallel()

	header := &HttpResponseHeader{
		MessageHeader: MessageHeader{Type: "http_response", ID: "resp-empty"},
		Status:        204,
		Headers:       map[string][]string{},
	}

	encoded, err := EncodeFrame(header, []byte{})
	if err != nil {
		t.Fatalf("EncodeFrame: %v", err)
	}

	f, err := DecodeFrame(encoded)
	if err != nil {
		t.Fatalf("DecodeFrame: %v", err)
	}

	// Zero-length body: remaining bytes after header are empty.
	if f.Body != nil {
		t.Errorf("expected nil body for zero-length body, got %d bytes", len(f.Body))
	}
}

func TestHeaderLength_AccurateSize(t *testing.T) {
	t.Parallel()

	header := &HttpRequestHeader{
		MessageHeader: MessageHeader{Type: "http_request", ID: "req-len"},
		Method:        "GET",
		Host:          "test.kagami.dev",
		Path:          "/",
		Query:         "",
		Headers:       map[string][]string{},
	}

	encoded, err := EncodeFrame(header, []byte("body"))
	if err != nil {
		t.Fatalf("EncodeFrame: %v", err)
	}

	headerLen := binary.BigEndian.Uint32(encoded[:4])

	// Marshal independently to verify.
	expectedJSON, err := json.Marshal(header)
	if err != nil {
		t.Fatalf("json.Marshal: %v", err)
	}

	if int(headerLen) != len(expectedJSON) {
		t.Errorf("header length = %d, want %d (actual JSON bytes)", headerLen, len(expectedJSON))
	}
}

func TestDecodeFrame_TooShort(t *testing.T) {
	t.Parallel()

	_, err := DecodeFrame([]byte{0, 0})
	if err == nil {
		t.Fatal("expected error for frame shorter than 4 bytes")
	}
}

func TestDecodeFrame_HeaderLenExceedsFrame(t *testing.T) {
	t.Parallel()

	// Header length says 100 but frame only has 4 bytes total (0 remaining).
	frame := make([]byte, 4)
	binary.BigEndian.PutUint32(frame, 100)

	_, err := DecodeFrame(frame)
	if err == nil {
		t.Fatal("expected error for header length exceeding frame")
	}
}

// --- Chunking tests ---

// decodeWireFrames is a test helper that decodes encoded wire frames
// returned by ChunkBody into Frame structs for inspection.
func decodeWireFrames(t *testing.T, wireFrames [][]byte) []Frame {
	t.Helper()
	frames := make([]Frame, len(wireFrames))
	for i, wf := range wireFrames {
		f, err := DecodeFrame(wf)
		if err != nil {
			t.Fatalf("DecodeFrame on wire frame %d: %v", i, err)
		}
		frames[i] = f
	}
	return frames
}

// extractBodies is a test helper that decodes wire frames and returns
// only the body slices, for use with ReassembleChunks.
func extractBodies(t *testing.T, wireFrames [][]byte) [][]byte {
	t.Helper()
	bodies := make([][]byte, len(wireFrames))
	for i, wf := range wireFrames {
		f, err := DecodeFrame(wf)
		if err != nil {
			t.Fatalf("DecodeFrame on wire frame %d: %v", i, err)
		}
		bodies[i] = f.Body
	}
	return bodies
}

func TestChunkBody_BelowThreshold(t *testing.T) {
	t.Parallel()

	header := &HttpRequestHeader{
		MessageHeader: MessageHeader{Type: "http_request", ID: "req-small"},
		Method:        "POST",
		Host:          "test.kagami.dev",
		Path:          "/upload",
		Query:         "",
		Headers:       map[string][]string{},
	}

	body := make([]byte, 100) // well under chunk size

	wireFrames, err := ChunkBody(header, body, ChunkSize)
	if err != nil {
		t.Fatalf("ChunkBody: %v", err)
	}

	if len(wireFrames) != 1 {
		t.Fatalf("expected 1 frame, got %d", len(wireFrames))
	}

	frames := decodeWireFrames(t, wireFrames)

	// Verify chunked field is NOT set (omitted from JSON).
	var decoded HttpRequestHeader
	if err := json.Unmarshal(frames[0].Header, &decoded); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if decoded.Chunked {
		t.Error("expected chunked to be false/omitted for small body")
	}

	if !bytes.Equal(frames[0].Body, body) {
		t.Error("body mismatch in single frame")
	}
}

func TestChunkBody_LargeBody_CorrectChunkCount(t *testing.T) {
	t.Parallel()

	chunkSize := 100
	bodySize := 350 // 100 + 100 + 100 + 50 = 4 chunks total
	body := make([]byte, bodySize)
	for i := range body {
		body[i] = byte(i % 256)
	}

	header := &HttpRequestHeader{
		MessageHeader: MessageHeader{Type: "http_request", ID: "req-big"},
		Method:        "POST",
		Host:          "test.kagami.dev",
		Path:          "/upload",
		Query:         "",
		Headers:       map[string][]string{},
	}

	wireFrames, err := ChunkBody(header, body, chunkSize)
	if err != nil {
		t.Fatalf("ChunkBody: %v", err)
	}

	// Expected: ceil(350/100) = 4 frames (1 initial + 3 continuation).
	expectedFrames := 4
	if len(wireFrames) != expectedFrames {
		t.Fatalf("expected %d frames, got %d", expectedFrames, len(wireFrames))
	}
}

func TestChunkBody_FirstFrame_ChunkedTrue(t *testing.T) {
	t.Parallel()

	chunkSize := 100
	body := make([]byte, 250)

	header := &HttpRequestHeader{
		MessageHeader: MessageHeader{Type: "http_request", ID: "req-chunk"},
		Method:        "POST",
		Host:          "test.kagami.dev",
		Path:          "/upload",
		Query:         "",
		Headers:       map[string][]string{},
	}

	wireFrames, err := ChunkBody(header, body, chunkSize)
	if err != nil {
		t.Fatalf("ChunkBody: %v", err)
	}

	frames := decodeWireFrames(t, wireFrames)

	// First frame should have chunked:true and be an http_request.
	var firstHeader HttpRequestHeader
	if err := json.Unmarshal(frames[0].Header, &firstHeader); err != nil {
		t.Fatalf("unmarshal first frame header: %v", err)
	}
	if !firstHeader.Chunked {
		t.Error("first frame should have chunked=true")
	}
	if firstHeader.Type != "http_request" {
		t.Errorf("first frame type = %q, want %q", firstHeader.Type, "http_request")
	}
	if len(frames[0].Body) != chunkSize {
		t.Errorf("first frame body = %d bytes, want %d", len(frames[0].Body), chunkSize)
	}
}

func TestChunkBody_ContinuationFrames(t *testing.T) {
	t.Parallel()

	chunkSize := 100
	body := make([]byte, 350)

	header := &HttpRequestHeader{
		MessageHeader: MessageHeader{Type: "http_request", ID: "req-cont"},
		Method:        "POST",
		Host:          "test.kagami.dev",
		Path:          "/upload",
		Query:         "",
		Headers:       map[string][]string{},
	}

	wireFrames, err := ChunkBody(header, body, chunkSize)
	if err != nil {
		t.Fatalf("ChunkBody: %v", err)
	}

	frames := decodeWireFrames(t, wireFrames)

	// frames[1], [2], [3] are continuation chunks.
	for i := 1; i < len(frames); i++ {
		var chunkHdr HttpBodyChunkHeader
		if err := json.Unmarshal(frames[i].Header, &chunkHdr); err != nil {
			t.Fatalf("unmarshal chunk %d header: %v", i, err)
		}
		if chunkHdr.Type != "http_body_chunk" {
			t.Errorf("chunk %d type = %q, want %q", i, chunkHdr.Type, "http_body_chunk")
		}
		if chunkHdr.ID != "req-cont" {
			t.Errorf("chunk %d id = %q, want %q", i, chunkHdr.ID, "req-cont")
		}

		isFinal := i == len(frames)-1
		if chunkHdr.Final != isFinal {
			t.Errorf("chunk %d final = %v, want %v", i, chunkHdr.Final, isFinal)
		}
	}
}

func TestChunkBody_Reassemble(t *testing.T) {
	t.Parallel()

	chunkSize := 100
	body := make([]byte, 350)
	for i := range body {
		body[i] = byte(i % 256)
	}

	header := &HttpRequestHeader{
		MessageHeader: MessageHeader{Type: "http_request", ID: "req-reasm"},
		Method:        "POST",
		Host:          "test.kagami.dev",
		Path:          "/upload",
		Query:         "",
		Headers:       map[string][]string{},
	}

	wireFrames, err := ChunkBody(header, body, chunkSize)
	if err != nil {
		t.Fatalf("ChunkBody: %v", err)
	}

	bodies := extractBodies(t, wireFrames)
	reassembled := ReassembleChunks(bodies)
	if !bytes.Equal(reassembled, body) {
		t.Error("reassembled body does not match original")
	}
}

func TestChunkBody_ExactMultiple(t *testing.T) {
	t.Parallel()

	chunkSize := 100
	body := make([]byte, 300) // exactly 3 chunks
	for i := range body {
		body[i] = byte(i % 256)
	}

	header := &HttpResponseHeader{
		MessageHeader: MessageHeader{Type: "http_response", ID: "resp-exact"},
		Status:        200,
		Headers:       map[string][]string{},
	}

	wireFrames, err := ChunkBody(header, body, chunkSize)
	if err != nil {
		t.Fatalf("ChunkBody: %v", err)
	}

	if len(wireFrames) != 3 {
		t.Fatalf("expected 3 frames, got %d", len(wireFrames))
	}

	bodies := extractBodies(t, wireFrames)
	reassembled := ReassembleChunks(bodies)
	if !bytes.Equal(reassembled, body) {
		t.Error("reassembled body does not match original")
	}
}

func TestChunkBody_ResponseHeader(t *testing.T) {
	t.Parallel()

	chunkSize := 50
	body := make([]byte, 120)

	header := &HttpResponseHeader{
		MessageHeader: MessageHeader{Type: "http_response", ID: "resp-chunk"},
		Status:        200,
		Headers:       map[string][]string{"Content-Type": {"text/plain"}},
	}

	wireFrames, err := ChunkBody(header, body, chunkSize)
	if err != nil {
		t.Fatalf("ChunkBody: %v", err)
	}

	frames := decodeWireFrames(t, wireFrames)

	// First frame is http_response with chunked:true.
	var first HttpResponseHeader
	if err := json.Unmarshal(frames[0].Header, &first); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if !first.Chunked {
		t.Error("first frame should have chunked=true")
	}
	if first.Type != "http_response" {
		t.Errorf("first frame type = %q, want %q", first.Type, "http_response")
	}
	if first.Status != 200 {
		t.Errorf("status = %d, want 200", first.Status)
	}
}

func TestChunkBody_DoesNotMutateCallerHeader(t *testing.T) {
	t.Parallel()

	header := &HttpRequestHeader{
		MessageHeader: MessageHeader{Type: "http_request", ID: "req-nomut"},
		Method:        "POST",
		Host:          "test.kagami.dev",
		Path:          "/upload",
		Query:         "",
		Headers:       map[string][]string{},
	}

	body := make([]byte, 250)
	_, err := ChunkBody(header, body, 100)
	if err != nil {
		t.Fatalf("ChunkBody: %v", err)
	}

	if header.Chunked {
		t.Error("ChunkBody mutated caller's header: Chunked should still be false")
	}
}

func TestParseHeaderType_ValidTypes(t *testing.T) {
	t.Parallel()

	for _, typ := range []string{"http_request", "http_response", "http_body_chunk", "ping", "pong", "error"} {
		hdr, _ := json.Marshal(MessageHeader{Type: typ, ID: "x"})
		got, err := ParseHeaderType(hdr)
		if err != nil {
			t.Errorf("ParseHeaderType(%q): %v", typ, err)
		}
		if got != typ {
			t.Errorf("ParseHeaderType(%q) = %q", typ, got)
		}
	}
}

func TestParseHeaderType_Unknown(t *testing.T) {
	t.Parallel()

	hdr := []byte(`{"type":"foo_bar","id":"x"}`)
	_, err := ParseHeaderType(hdr)
	if err == nil {
		t.Fatal("expected error for unknown type")
	}
}

func TestParseHeaderType_MissingType(t *testing.T) {
	t.Parallel()

	hdr := []byte(`{"id":"x"}`)
	_, err := ParseHeaderType(hdr)
	if err == nil {
		t.Fatal("expected error for missing type")
	}
}

func TestParseHeaderType_InvalidJSON(t *testing.T) {
	t.Parallel()

	_, err := ParseHeaderType([]byte(`{broken`))
	if err == nil {
		t.Fatal("expected error for invalid JSON")
	}
}

func TestChunkBody_WithDefaultChunkSize(t *testing.T) {
	t.Parallel()

	// Create a body larger than ChunkSize (512KB).
	body := make([]byte, ChunkSize+1000)
	for i := range body {
		body[i] = byte(i % 256)
	}

	header := &HttpRequestHeader{
		MessageHeader: MessageHeader{Type: "http_request", ID: "req-default"},
		Method:        "POST",
		Host:          "test.kagami.dev",
		Path:          "/big",
		Query:         "",
		Headers:       map[string][]string{},
	}

	wireFrames, err := ChunkBody(header, body, ChunkSize)
	if err != nil {
		t.Fatalf("ChunkBody: %v", err)
	}

	if len(wireFrames) != 2 {
		t.Fatalf("expected 2 frames, got %d", len(wireFrames))
	}

	bodies := extractBodies(t, wireFrames)
	reassembled := ReassembleChunks(bodies)
	if !bytes.Equal(reassembled, body) {
		t.Error("reassembled body does not match original")
	}
}

func TestChunkBody_BoundaryExact(t *testing.T) {
	t.Parallel()

	// Body exactly at chunk size -- should produce a single frame, no chunking.
	body := make([]byte, 100)

	header := &HttpRequestHeader{
		MessageHeader: MessageHeader{Type: "http_request", ID: "req-boundary"},
		Method:        "POST",
		Host:          "test.kagami.dev",
		Path:          "/",
		Query:         "",
		Headers:       map[string][]string{},
	}

	wireFrames, err := ChunkBody(header, body, 100)
	if err != nil {
		t.Fatalf("ChunkBody: %v", err)
	}

	if len(wireFrames) != 1 {
		t.Fatalf("expected 1 frame for body == chunkSize, got %d", len(wireFrames))
	}

	frames := decodeWireFrames(t, wireFrames)
	var decoded HttpRequestHeader
	if err := json.Unmarshal(frames[0].Header, &decoded); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if decoded.Chunked {
		t.Error("body exactly at chunk size should not be chunked")
	}
}

File: packages/kagami/src/protocol.test.ts
import { describe, it, expect } from "vitest";
import {
  encodeFrame,
  decodeFrame,
  encodeChunked,
  reassembleChunks,
  DEFAULT_CHUNK_SIZE,
  type HttpRequestHeader,
  type HttpResponseHeader,
  type PingHeader,
  type PongHeader,
  type HttpBodyChunkHeader,
} from "./protocol.js";

// --- Helpers ---

/** Create a Uint8Array filled with a repeating byte pattern of the given length. */
function makeBody(length: number, seed: number = 0xab): Uint8Array {
  const buf = new Uint8Array(length);
  for (let i = 0; i < length; i++) {
    buf[i] = (seed + i) & 0xff;
  }
  return buf;
}

/** Build a minimal HttpRequestHeader for testing. */
function makeRequestHeader(
  overrides?: Partial<HttpRequestHeader>,
): HttpRequestHeader {
  return {
    type: "http_request",
    id: "req-001",
    method: "GET",
    host: "api.my-homelab.kagami.example.com",
    path: "/users",
    query: "page=1",
    headers: { "content-type": ["application/json"] },
    ...overrides,
  };
}

/** Build a minimal HttpResponseHeader for testing. */
function makeResponseHeader(
  overrides?: Partial<HttpResponseHeader>,
): HttpResponseHeader {
  return {
    type: "http_response",
    id: "req-001",
    status: 200,
    headers: { "content-type": ["application/json"] },
    ...overrides,
  };
}

// --- encodeFrame / decodeFrame ---

describe("encodeFrame + decodeFrame", () => {
  it("encoding HttpRequestHeader + body produces a valid binary frame", () => {
    const header = makeRequestHeader();
    const body = new TextEncoder().encode('{"hello":"world"}');

    const frame = encodeFrame(header, body);

    // Frame is an ArrayBuffer
    expect(frame).toBeInstanceOf(ArrayBuffer);

    // First 4 bytes are the header length (uint32 BE)
    const view = new DataView(frame);
    const headerLen = view.getUint32(0, false);
    expect(headerLen).toBeGreaterThan(0);

    // Remaining bytes after header should equal body length
    expect(frame.byteLength).toBe(4 + headerLen + body.byteLength);

    // JSON header should be valid and match input
    const headerBytes = new Uint8Array(frame, 4, headerLen);
    const parsedHeader = JSON.parse(new TextDecoder().decode(headerBytes));
    expect(parsedHeader.type).toBe("http_request");
    expect(parsedHeader.id).toBe("req-001");
    expect(parsedHeader.method).toBe("GET");

    // Body bytes should match
    const bodyBytes = new Uint8Array(frame, 4 + headerLen);
    expect(bodyBytes).toEqual(body);
  });

  it("decoding a binary frame produces correct HttpResponseHeader and raw body bytes", () => {
    const header = makeResponseHeader({ status: 201 });
    const body = new TextEncoder().encode("response body");

    const frame = encodeFrame(header, body);
    const decoded = decodeFrame(frame);

    expect(decoded.header.type).toBe("http_response");
    expect(decoded.header.id).toBe("req-001");
    const respHeader = decoded.header as HttpResponseHeader;
    expect(respHeader.status).toBe(201);
    expect(respHeader.headers["content-type"]).toEqual(["application/json"]);
    expect(new TextDecoder().decode(decoded.body)).toBe("response body");
  });

  it("round-trips binary body data through encode/decode preserving exact bytes", () => {
    const header = makeRequestHeader({ id: "bin-001" });
    const body = makeBody(1024);

    const frame = encodeFrame(header, body);
    const decoded = decodeFrame(frame);

    expect(decoded.body).toEqual(body);
  });

  it("rejects frames with missing 'type' field in JSON header", () => {
    // Manually build a frame with a header missing `type`
    const headerJson = JSON.stringify({ id: "no-type" });
    const headerBytes = new TextEncoder().encode(headerJson);
    const frame = new ArrayBuffer(4 + headerBytes.byteLength);
    new DataView(frame).setUint32(0, headerBytes.byteLength, false);
    new Uint8Array(frame).set(headerBytes, 4);

    expect(() => decodeFrame(frame)).toThrow(/type/i);
  });

  it("rejects frames with missing 'id' field in JSON header", () => {
    const headerJson = JSON.stringify({ type: "ping" });
    const headerBytes = new TextEncoder().encode(headerJson);
    const frame = new ArrayBuffer(4 + headerBytes.byteLength);
    new DataView(frame).setUint32(0, headerBytes.byteLength, false);
    new Uint8Array(frame).set(headerBytes, 4);

    expect(() => decodeFrame(frame)).toThrow(/id/i);
  });

  it("rejects frames where header length exceeds frame size", () => {
    // Create a frame that claims a header length larger than the actual data
    const frame = new ArrayBuffer(8); // only 4 bytes of "data" after length
    new DataView(frame).setUint32(0, 100, false); // claims 100 bytes of header

    expect(() => decodeFrame(frame)).toThrow(/exceeds frame size/);
  });

  it("handles frame with zero-length body", () => {
    const header = makeRequestHeader({ id: "empty-body" });
    const frame = encodeFrame(header); // no body argument

    const decoded = decodeFrame(frame);
    expect(decoded.header.type).toBe("http_request");
    expect(decoded.body.byteLength).toBe(0);
  });

  it("control messages (ping) round-trip with no body", () => {
    const header: PingHeader = { type: "ping", id: "ping-001" };
    const frame = encodeFrame(header);
    const decoded = decodeFrame(frame);

    expect(decoded.header.type).toBe("ping");
    expect(decoded.header.id).toBe("ping-001");
    expect(decoded.body.byteLength).toBe(0);
  });

  it("control messages (pong) round-trip with no body", () => {
    const header: PongHeader = { type: "pong", id: "pong-001" };
    const frame = encodeFrame(header);
    const decoded = decodeFrame(frame);

    expect(decoded.header.type).toBe("pong");
    expect(decoded.header.id).toBe("pong-001");
    expect(decoded.body.byteLength).toBe(0);
  });

  it("header length field accurately reflects the JSON header byte length", () => {
    const header = makeRequestHeader({
      id: "len-check",
      headers: { "x-custom": ["value-with-utf8-\u00e9"] },
    });
    const frame = encodeFrame(header);
    const view = new DataView(frame);
    const headerLen = view.getUint32(0, false);

    const expectedBytes = new TextEncoder().encode(JSON.stringify(header));
    expect(headerLen).toBe(expectedBytes.byteLength);
  });
});

// --- Chunking ---

describe("encodeChunked", () => {
  it("small body produces a single frame with no chunked field", () => {
    const header = makeResponseHeader({ id: "small-001" });
    const body = makeBody(100);

    const { frames } = encodeChunked(header, body);

    expect(frames).toHaveLength(1);

    // Decode and verify no chunked flag
    const decoded = decodeFrame(frames[0]);
    const respHeader = decoded.header as HttpResponseHeader;
    expect(respHeader.chunked).toBeUndefined();
    expect(decoded.body).toEqual(body);
  });

  it("body exactly at chunk size produces a single frame", () => {
    const chunkSize = 256;
    const header = makeResponseHeader({ id: "exact-001" });
    const body = makeBody(chunkSize);

    const { frames } = encodeChunked(header, body, chunkSize);

    expect(frames).toHaveLength(1);
    const decoded = decodeFrame(frames[0]);
    expect((decoded.header as HttpResponseHeader).chunked).toBeUndefined();
    expect(decoded.body).toEqual(body);
  });

  it("large body is split into correct number of chunks", () => {
    const chunkSize = 256;
    const bodySize = 700; // 256 + 256 + 188 = 3 frames
    const header = makeResponseHeader({ id: "chunk-001" });
    const body = makeBody(bodySize);

    const { frames } = encodeChunked(header, body, chunkSize);

    // 1 initial + 2 continuation = 3 frames
    expect(frames).toHaveLength(3);

    // First frame: http_response with chunked: true
    const first = decodeFrame(frames[0]);
    expect(first.header.type).toBe("http_response");
    expect((first.header as HttpResponseHeader).chunked).toBe(true);
    expect(first.body.byteLength).toBe(chunkSize);

    // Second frame: http_body_chunk with final: false
    const second = decodeFrame(frames[1]);
    expect(second.header.type).toBe("http_body_chunk");
    expect((second.header as HttpBodyChunkHeader).final).toBe(false);
    expect(second.body.byteLength).toBe(chunkSize);

    // Third frame: http_body_chunk with final: true
    const third = decodeFrame(frames[2]);
    expect(third.header.type).toBe("http_body_chunk");
    expect((third.header as HttpBodyChunkHeader).final).toBe(true);
    expect(third.body.byteLength).toBe(bodySize - 2 * chunkSize);
  });

  it("continuation frames share the same id as the initial frame", () => {
    const chunkSize = 100;
    const header = makeRequestHeader({ id: "shared-id-001" });
    const body = makeBody(250);

    const { frames } = encodeChunked(header, body, chunkSize);

    for (const frame of frames) {
      const decoded = decodeFrame(frame);
      expect(decoded.header.id).toBe("shared-id-001");
    }
  });

  it("uses DEFAULT_CHUNK_SIZE when no chunkSize is specified", () => {
    expect(DEFAULT_CHUNK_SIZE).toBe(524288);

    const header = makeResponseHeader({ id: "default-chunk" });
    // Body just over 512KB
    const body = makeBody(DEFAULT_CHUNK_SIZE + 100);

    const { frames } = encodeChunked(header, body);

    // Should produce 2 frames: initial (512KB) + final (100 bytes)
    expect(frames).toHaveLength(2);
  });
});

describe("reassembleChunks", () => {
  it("reassembling all chunks produces the original body exactly", () => {
    const chunkSize = 256;
    const body = makeBody(700);
    const header = makeResponseHeader({ id: "reassemble-001" });

    const { frames } = encodeChunked(header, body, chunkSize);

    // Extract body from each frame
    const chunks: Uint8Array[] = frames.map((f) => decodeFrame(f).body);

    const reassembled = reassembleChunks(chunks);
    expect(reassembled).toEqual(body);
  });

  it("reassembling a single chunk returns the original body", () => {
    const body = makeBody(100);
    const reassembled = reassembleChunks([body]);
    expect(reassembled).toEqual(body);
  });

  it("reassembling zero chunks returns empty Uint8Array", () => {
    const reassembled = reassembleChunks([]);
    expect(reassembled.byteLength).toBe(0);
  });

  it("reassembles a large body (multiple of chunk size) correctly", () => {
    const chunkSize = 200;
    const body = makeBody(600); // exactly 3 chunks
    const header = makeResponseHeader({ id: "exact-multi" });

    const { frames } = encodeChunked(header, body, chunkSize);
    expect(frames).toHaveLength(3);

    const chunks = frames.map((f) => decodeFrame(f).body);
    const reassembled = reassembleChunks(chunks);
    expect(reassembled).toEqual(body);
  });

  it("round-trips a body through chunking and reassembly with binary data", () => {
    const chunkSize = 128;
    // Generate a body with all possible byte values
    const body = new Uint8Array(256);
    for (let i = 0; i < 256; i++) {
      body[i] = i;
    }

    const header = makeRequestHeader({ id: "binary-roundtrip" });
    const { frames } = encodeChunked(header, body, chunkSize);

    const chunks = frames.map((f) => decodeFrame(f).body);
    const reassembled = reassembleChunks(chunks);

    expect(reassembled).toEqual(body);
  });
});
